File: backend/controller/petugasLain_controller.go
package controller

import (
	"TaskPus/database"
	"TaskPus/helpers"
	"TaskPus/model"
	"TaskPus/structs"
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

func GetPetugasLain(c *gin.Context) {
	var petugas []model.PetugasLain
	var response []structs.PetugasLainResponse

	if err := database.DB.Find(&petugas).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to fetch petugas lain",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	for _, u := range petugas {
		r := structs.PetugasLainResponse{
			Id:      u.Id,
			Nama:    u.Nama,
			NIP:     u.NIP,
			Jabatan: u.Jabatan,
			Pangkat: u.Pangkat,
			Unit:    u.Unit,
		}
		response = append(response, r)
	}

	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "List Data Petugas Lain",
		Data:    response,
	})
}

func CreatePetugasLain(c *gin.Context) {
	var req = structs.CreatePetugasLainRequest{}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusUnprocessableEntity, structs.ErrorResponse{
			Success: false,
			Message: "Validation Errors",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	petugas := model.PetugasLain{
		Nama:    req.Nama,
		NIP:     req.NIP,
		Jabatan: req.Jabatan,
		Pangkat: req.Pangkat,
		Unit:    req.Unit,
	}

	if err := database.DB.Create(&petugas).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to create petugas lain",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	r := structs.PetugasLainResponse{
		Id:      petugas.Id,
		Nama:    petugas.Nama,
		NIP:     petugas.NIP,
		Jabatan: petugas.Jabatan,
		Pangkat: petugas.Pangkat,
		Unit:    petugas.Unit,
	}

	c.JSON(http.StatusCreated, structs.SuccessResponse{
		Success: true,
		Message: "Petugas Lain created successfully",
		Data:    r,
	})
}

func GetPetugasLainById(c *gin.Context) {
	id := c.Param("id")

	var petugas model.PetugasLain

	if err := database.DB.First(&petugas, id).Error; err != nil {
		c.JSON(http.StatusNotFound, structs.ErrorResponse{
			Success: false,
			Message: "User not found",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	r := structs.PetugasLainResponse{
		Id:      petugas.Id,
		Nama:    petugas.Nama,
		NIP:     petugas.NIP,
		Jabatan: petugas.Jabatan,
		Pangkat: petugas.Pangkat,
		Unit:    petugas.Unit,
	}

	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "Petugas Lain found",
		Data:    r,
	})
}

func UpdatePetugasLain(c *gin.Context) {
	id := c.Param("id")

	var petugas model.PetugasLain
	if err := database.DB.First(&petugas, id).Error; err != nil {
		c.JSON(http.StatusNotFound, structs.ErrorResponse{
			Success: false,
			Message: "Petugas Lain not found",
		})
		return
	}

	var req structs.UpdatePetugasLainRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusUnprocessableEntity, structs.ErrorResponse{
			Success: false,
			Message: "Validation Errors",
			Errors:  helpers.TranslateErrorMessage(err),
		})
		return
	}

	// Update field umum
	petugas.Nama = req.Nama
	petugas.NIP = req.NIP
	petugas.Jabatan = req.Jabatan
	petugas.Pangkat = req.Pangkat
	petugas.Unit = req.Unit

	if err := database.DB.Session(&gorm.Session{FullSaveAssociations: true}).Save(&petugas).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to update petugas lain",
		})
		return
	}

	// Response
	r := structs.PetugasLainResponse{
		Id:      petugas.Id,
		Nama:    petugas.Nama,
		NIP:     petugas.NIP,
		Jabatan: petugas.Jabatan,
		Pangkat: petugas.Pangkat,
		Unit:    petugas.Unit,
	}

	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "Pegawai updated successfully",
		Data:    r,
	})
}

func DeletePetugasLain(c *gin.Context) {
	id := c.Param("id")

	var petugas model.PetugasLain

	if err := database.DB.First(&petugas, id).Error; err != nil {
		c.JSON(http.StatusNotFound, structs.ErrorResponse{
			Success: false,
			Message: "Petugas Lain not found",
		})
		return
	}

	if err := database.DB.Delete(&petugas).Error; err != nil {
		c.JSON(http.StatusInternalServerError, structs.ErrorResponse{
			Success: false,
			Message: "Failed to delete petugas lain",
		})
		return
	}

	c.JSON(http.StatusOK, structs.SuccessResponse{
		Success: true,
		Message: "Pegawai deleted successfully",
	})
}

File: backend/app/Http/Controllers/PetugasLainController.php
<?php

namespace App\Http\Controllers;

use App\Models\Petugas_Lain;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Hash;
use Illuminate\Validation\ValidationException;
use Illuminate\Support\Facades\DB;
use Tymon\JWTAuth\Facades\JWTAuth;
use function App\Helpers\successResponse;
use function App\Helpers\errorResponse;

class PetugasLainController extends Controller
{
    //
    public function GetPetugas_Lain()
    {
        $petugas_Lain = Petugas_Lain::get();

        $response = $petugas_Lain->map(function ($u) {
            return [
                'id'         => $u->id,
                'nama'       => $u->nama,
                'nip'       => $u->nip,
                'jabatan'       => $u->jabatan,
                'pangkat'       => $u->pangkat,
                'unit'      => $u->unit,
            ];
        });

        return response()->json([
            'success' => true,
            'message' => 'List Data Petugas_Lain',
            'data'    => $response
        ]);
    }

    public function CreatePetugas_Lain(Request $request)
    {
        $data = $request->validate([
            'nama'     => 'required|string',
            'nip'     => 'required|string',
            'jabatan'     => 'required|string',
            'pangkat'     => 'required|string',
            'unit'     => 'required|string',
        ]);

        $petugas_Lain = Petugas_Lain::create($data);

        return successResponse('Petugas_Lain created successfully', [
            'id'         => $petugas_Lain->id,
            'nama'       => $petugas_Lain->nama,
            'nip'   => $petugas_Lain->nip,
            'jabatan'   => $petugas_Lain->jabatan,
            'pangkat'   => $petugas_Lain->pangkat,
            'unit'   => $petugas_Lain->unit,
        ]);
    }

    public function GetPetugas_LainById($id)
    {
        $petugas_Lain = Petugas_Lain::findOrFail($id);

        return response()->json([
            'success' => true,
            'message' => 'Petugas_Lain found',
            'data'    => $this->Petugas_LainResponse($petugas_Lain)
        ]);
    }

    public function UpdatePetugas_Lain(Request $request, $id)
    {
        $petugas_Lain = Petugas_Lain::findOrFail($id);

        $data = $request->validate([
            'name'     => 'sometimes|required|string' . $petugas_Lain->id,
            'nip'     => 'sometimes|required|string' . $petugas_Lain->id,
            'jabatan'     => 'sometimes|required|string' . $petugas_Lain->id,
            'pangkat'     => 'sometimes|required|string' . $petugas_Lain->id,
            'unit'     => 'sometimes|required|string' . $petugas_Lain->id,
        ]);

        $petugas_Lain->update($data);

        return successResponse('Petugas_Lain updated successfully', $this->Petugas_LainResponse($petugas_Lain));
    }

    public function DeletePetugas_Lain($id)
    {
        $petugas_Lain = Petugas_Lain::findOrFail($id);
        $petugas_Lain->delete();

        return successResponse('Petugas_Lain deleted successfully');
    }

    private function Petugas_LainResponse(Petugas_Lain $petugas_Lain)
    {
        return [
            'id'         => $petugas_Lain->id,
            'nama'       => $petugas_Lain->nama,
            'nip'   => $petugas_Lain->nip,
            'jabatan'   => $petugas_Lain->jabatan,
            'pangkat'   => $petugas_Lain->pangkat,
            'unit'   => $petugas_Lain->unit,
        ];
    }
}
